### PR TITLE
added basic disambiguation of same name files

### LINKED ISSUE #551: Disambiguate files with the same name
Currently, if opening two files with the same name (but obviously a different location) there is nothing to tell which one is which if you are not keeping track of their position.  
VSCode does this by adding part of their path to their titles, 
image

### PR DESCRIPTION
**Pull request by nheuillet**
This PR introduces some visual syntactic sugar around same name files from different folders.

* Every path created is truncated to 20 characters, which is completely arbitrary and can be discussed/changed*
- When two same-name files are open, a path hint will appear for as long as multiple same name files are opened
- When two files from the same workspace are open, the names of the folders (full path minus workspace root) are used
- When a file from outside of the workspace is opened, the full path is used
- When a file at the workspace root is opened, `./` is used

The implementation is not as efficient as it could be (some clones and a lot of conversions), but I am no Rust expert, so feel free to show me how to improve :)

closes #551

### CODE REVIEW

File: lapce-ui/src/editor/tab_header_content.rs [hunk 1]
```diff
@@ -248,6 +249,79 @@ impl LapceEditorTabHeaderContent {
...
+
+    fn get_effective_path(
+        &mut self,
+        workspace_path: Option<PathBuf>,
+        file_path: &PathBuf,
+    ) -> String {
+        let workspace_path = match workspace_path {
+            Some(p) => p,
+            None => {
+                let mut file_path_str = match file_path.to_str() {
```
Comment (MinusGix) [inline lapce-ui/src/editor/tab_header_content.rs:261]: Minor: `file_path.to_str().unwrap_or("")` (then call `.to_string` on that)

Comment (MinusGix) [reply lapce-ui/src/editor/tab_header_content.rs:261]: You could also just use https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.to_string_lossy

Comment (MinusGix) [reply lapce-ui/src/editor/tab_header_content.rs:261]: Minor: In Rust, shadowing names (so using file_path for the String version too) is normal when they are not needed anymore (like your converting it into a String).

File: lapce-ui/src/editor/tab_header_content.rs [hunk 2]
```diff
@@ -248,6 +249,79 @@ impl LapceEditorTabHeaderContent {
...
+                    Some(p) => p.to_string(),
+                    None => return "".to_string(),
+                };
+
+                file_path_str.truncate(20);
```
Comment (MinusGix) [inline lapce-ui/src/editor/tab_header_content.rs:266]: This has potential issue if the name contains special characters, since 20 might lie in the middle of a multibyte character which would panic.  
I add some issue in finding if there was a better solution than truncate for this task, unfortunately.  
```rust
let file_path_str = file_path_str.char_indices()
        .filter(|(i, _)| *i < 20)
        .map(|(_, ch)| ch)
        .collect::<String>();
```
Ugly, and I feel like there should be a better way, but I believe it should work.

File: lapce-ui/src/editor/tab_header_content.rs [hunk 3]
```diff
@@ -248,6 +249,79 @@ impl LapceEditorTabHeaderContent {
...
+        // file is a workspace file: we can truncate to workspace folders
+        if file_path.starts_with(&workspace_path) {
+            let mut reversed_path: Vec<String> = Vec::<String>::new();
+            // let mut parent = file_path.parent();
+            let mut file_path_mut = file_path.clone();
+
+            // We dont need to keep the file name, only the parents.
+            file_path_mut.pop();
+
+            while file_path_mut.cmp(&workspace_path) != Ordering::Equal {
```
Comment (MinusGix) [inline lapce-ui/src/editor/tab_header_content.rs:281]: I think this can just be `while file_path_mut != workspace_path {`?

File: lapce-ui/src/editor/tab_header_content.rs (hunk 2 above)
Comment (nheuillet) [reply lapce-ui/src/editor/tab_header_content.rs:266]: It looks like it works like a charm yeah! I'll add to test for unicode-based names to avoid regressions.

File: lapce-ui/src/editor/tab_header_content.rs [hunk 4]
```diff
@@ -351,9 +425,31 @@ impl Widget<LapceTabData> for LapceEditorTabHeaderContent {
...
+            match child {
+                EditorTabChild::Editor(view_id, _, _) => {
+                    let editor = data.main_split.editors.get(view_id).unwrap();
+                    if let BufferContent::File(path) = &editor.content {
+                        if let Some(file_name) = path.file_name() {
+                            let name = file_name.to_str().unwrap();
+                            let nb = match name_count.get(name) {
+                                Some(name) => *name,
+                                None => 0,
+                            };
```
Comment (MinusGix) [inline lapce-ui/src/editor/tab_header_content.rs:439]: Minor: `name_count.get(name).copied().unwrap_or(0)`, which copies the held reference to an integer value and then unwraps to it or 0.

File: lapce-ui/src/editor/tab_header_content.rs [hunk 5]
```diff
@@ -362,6 +458,18 @@ impl Widget<LapceTabData> for LapceEditorTabHeaderContent {
...
+                                let nb = name_count
+                                    .get(file_name.to_str().unwrap())
+                                    .unwrap();
+                                if *nb > 1 {
+                                    file_path = format!(
+                                        " {}",
+                                        self.get_effective_path(
+                                            data.workspace.path.clone(),
+                                            path
+                                        )
```
Comment (MinusGix) [inline lapce-ui/src/editor/tab_header_content.rs:470]: `get_effective_path` could take an `Option<&PathBuf>` to avoid unneeded cloning. (`data.workspace.path.as_ref()` to get a reference into the option)

**Comment by MinusGix** [comment]
The file-counting logic should maybe go in `update` instead of `layout`, and it should be a field on data somewhere so that it can be reused. However, I think we can avoid having the hashmap and string allocations at all.  
Most uses of the editor would have significantly less than even 100 editors open and so just iterating over each of them to check if they have a duplicate filename seems like it would be better (compared to allocating a bunch of strings).  
Ex:  
```rust
for (i, child) in editor_tab.children.iter().enumerate() {
    // the variables defined at the start
    match child {
        EditorTabChild::Editor(view_id, _, _) => {
            let editor = data.main_split.editors.get(view_id).unwrap();
            if let Some(file_name) = path.file_name() {
                if has_other_editors_with_same_name(data, editor_tab, i, file_name) {
                    // though, file_path could perhaps be an Option<String>?
                    file_path = format!(" {}", self.get_effective_path(data.workspace.path.clone()));
                }
            }
        }
    }
}

// define some separate helper function if you want
/// Checks if there is another editor with the same file name
fn has_other_editors_with_same_name(data: &LapceTabData, editor_tab: &LapceEditorTabData, index: usize, file_name: &OsStr) -> bool {
    for (i, child) in editor_tab.children.iter().enumerate() {
        // Skip past itself
        if i == index {
            continue;
        }
        
        match child {
            EditorTabChild::Editor(view_id, _, _) => {
                let editor = data.main_split.editors.get(view_id).unwrap();
                if let BufferContent::File(path) = &editor.content {
                    if let Some(child_filename) = path.file_name() {
                        if child_filename == file_name {
                            // There is a copy
                            return true;
                        }
                    }
                }
            }
            EditorTabChild::Settings(_, _) => {}
        }
    }
    
    // There was no copy
    false
}
```
(Note: The above is untested, I wrote it out from looking at the existing code)

File: lapce-ui/src/editor/tab_header_content.rs [hunk 6]
```diff
@@ -248,6 +249,79 @@ impl LapceEditorTabHeaderContent {
...
+        file_path: &PathBuf,
+    ) -> String {
+        let workspace_path = match workspace_path {
+            Some(p) => p,
+            None => {
+                // file_path.truncate(20);
+                let file_path = file_path
+                    .to_string_lossy()
+                    .char_indices()
+                    .filter(|(i, _)| *i < 20)
```
Comment (bugadani) [inline lapce-ui/src/editor/tab_header_content.rs:265]: Are you looking for `.chars().take(20)`? :)

File: lapce-ui/src/editor/tab_header_content.rs [hunk 7]
```diff
@@ -248,6 +249,79 @@ impl LapceEditorTabHeaderContent {
...
+                let file_name_str = file_name.to_str().unwrap().to_string();
+
+                reversed_path.push(file_name_str);
+                if !file_path_mut.pop() {
+                    break;
+                }
+            }
+
+            // File is at workspace root
+            if reversed_path.len() == 0 {
```
Comment (bugadani) [inline lapce-ui/src/editor/tab_header_content.rs:292]: Clippy will complain here, you can use `.is_empty()` instead

File: lapce-ui/src/editor/tab_header_content.rs [hunk 8]
```diff
@@ -248,6 +249,79 @@ impl LapceEditorTabHeaderContent {
             ));
         }
     }
+
+    fn get_effective_path(
+        &mut self,
+        workspace_path: Option<PathBuf>,
+        file_path: &PathBuf,
```
Comment (bugadani) [inline lapce-ui/src/editor/tab_header_content.rs:256]: Taking a `&Path` would be more idiomatic.

File: lapce-ui/src/editor/tab_header_content.rs [hunk 9]
```diff
@@ -248,6 +249,79 @@ impl LapceEditorTabHeaderContent {
...
+                    .char_indices()
+                    .filter(|(i, _)| *i < 20)
+                    .map(|(_, ch)| ch)
+                    .collect::<String>();
+                return file_path;
+            }
+        };
+
+        // file is a workspace file: we can truncate to workspace folders
+        if file_path.starts_with(&workspace_path) {
```
Comment (bugadani) [inline lapce-ui/src/editor/tab_header_content.rs:273]: I haven't given this much thought, really, but if you want to trim the workspace prefix, you can do that with `str`'s [`strip_prefix`](https://doc.rust-lang.org/std/primitive.str.html#method.strip_prefix). I think the problems panel implementation does this already if you want an example.

File: lapce-ui/src/editor/tab_header_content.rs [hunk 10]
```diff
@@ -351,9 +425,31 @@ impl Widget<LapceTabData> for LapceEditorTabHeaderContent {
 
         self.rects.clear();
         let mut x = 0.0;
+        let mut name_count: HashMap<String, i32> = HashMap::new();
+        for (_i, child) in editor_tab.children.iter().enumerate() {
```
Comment (bugadani) [inline lapce-ui/src/editor/tab_header_content.rs:429]: Why do you `enumerate` if you ignore the index afterwards?

File: lapce-ui/src/editor/tab_header_content.rs (hunk 4 above)
Comment (bugadani) [reply lapce-ui/src/editor/tab_header_content.rs:439]: I think we want to use the entry API here. That would avoid hashing the name twice.

Also, "the number equals the name" is hella confusing :) Using the same variable name to refer to a different thing basically one line away isn't very clear.

File: lapce-ui/src/editor/tab_header_content.rs [hunk 11]
```diff
@@ -351,9 +425,31 @@ impl Widget<LapceTabData> for LapceEditorTabHeaderContent {
...
+            match child {
```
Comment (bugadani) [inline lapce-ui/src/editor/tab_header_content.rs:430]: This would be better as an `if let` since you only handle a single branch.

File: lapce-ui/src/editor/tab_header_content.rs [hunk 12]
```diff
@@ -384,7 +492,19 @@ impl Widget<LapceTabData> for LapceEditorTabHeaderContent {
                 )
                 .build()
                 .unwrap();
-            let text_size = text_layout.size();
+            let path_layout = ctx
```
Comment (bugadani) [inline lapce-ui/src/editor/tab_header_content.rs:495]: I don't see clearly on Github's UI, but you might want to check if `file_path` is empty here and skip this whole thing in that case. `TabRect` could have an optional `path_layout` to make it work.

File: lapce-ui/src/editor/tab_header_content.rs [hunk 13]
```diff
@@ -362,6 +458,18 @@ impl Widget<LapceTabData> for LapceEditorTabHeaderContent {
                         if let Some(file_name) = path.file_name() {
                             if let Some(s) = file_name.to_str() {
                                 text = s.to_string();
+                                let nb = name_count
+                                    .get(file_name.to_str().unwrap())
```
Comment (bugadani) [inline lapce-ui/src/editor/tab_header_content.rs:462]: `s` is already this value, isn't is?

File: lapce-ui/src/editor/tab_header_content.rs [hunk 14]
```diff
@@ -351,9 +425,31 @@ impl Widget<LapceTabData> for LapceEditorTabHeaderContent {
 
         self.rects.clear();
         let mut x = 0.0;
+        let mut name_count: HashMap<String, i32> = HashMap::new();
```
Comment (bugadani) [inline lapce-ui/src/editor/tab_header_content.rs:428]: I think building this hashmap could have a better place in `update`. In general, you want to do it when the data changes, not when the UI changes.

**Comment by bugadani** [comment]
> The file-counting logic should maybe go in update instead of layout

I should have read the comments before adding my own. Sorry about that :')

File: lapce-ui/src/editor/tab_header_content.rs (hunk 6 above)
Comment (MinusGix) [reply lapce-ui/src/editor/tab_header_content.rs:265]: Yeah that's better, forgot about `take` when I suggested it.

File: lapce-ui/src/editor/tab_header_content.rs (hunk 9 above)
Comment (nheuillet) [reply lapce-ui/src/editor/tab_header_content.rs:273]: I do want to trim the workspace prefix (make it so that the path shown will be relative inside the workspace). Nevertheless, using `strip_prefix` involves some normalization issues (trailing slashes when popping the file, etc.), so I would say using pathbuff would be the easiest way to handle this case. There might be a more efficient way though, and I will try to look into it

**Comment by bugadani** [comment]
Wait, I have reviewed this PR already? Can we move forward on this?

**Comment by nheuillet** [comment]
@bugadani there was some comments about the way to store and handle / compute the data that I never got around to resolve in the end. If someone wants to take care of it then by all means. Otherwise, I'll revisit and complete this PR in a few weeks, as I want to finish PSP first.

**Comment by hbina** [comment]
Let me give it a shot.

**Comment by nheuillet** [comment]
> Let me give it a shot.

Gladly! thanks a lot @hbina

**Comment by hbina** [comment]
Do i need to create a new PR or can I reuse this somehow?

**Comment by panekj** [comment]
> Do i need to create a new PR or can I reuse this somehow?

create new PR

**Comment by nheuillet** [comment]
@hbina Create a new PR but feel free to copy whatever you need from this PR :)

**Comment by nheuillet** [comment]
closing as #1521 got merged
